src/backend/distributed/catalog/distobject.c

```diff
@@ -0,0 +1,336 @@
+ */
+void
+unmarkObjectDistributed(const ObjectAddress *address)
+{
+	Relation pgDistObjectRel = NULL;
+	ScanKeyData key[2] = { 0 };
+	SysScanDesc pgDistObjectScan = NULL;
+	HeapTuple pgDistObjectTup = NULL;
+
+	pgDistObjectRel = heap_open(DistObjectRelationId(), RowExclusiveLock);
+
+	/* scan pg_dist_object for classid = $1 AND objid = $2 using an index */
+	ScanKeyInit(&key[0], Anum_pg_dist_object_classid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(address->classId));
+	ScanKeyInit(&key[1], Anum_pg_dist_object_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(address->objectId));
+	pgDistObjectScan = systable_beginscan(pgDistObjectRel,
+										  DistObjectClassIDObjectIDIndexId(), true, NULL,
+										  2, key);
+
+	while (HeapTupleIsValid(pgDistObjectTup = systable_getnext(pgDistObjectScan)))
+	{
+		CatalogTupleDelete(pgDistObjectRel, &pgDistObjectTup->t_self);
+	}
+
+	systable_endscan(pgDistObjectScan);
+	relation_close(pgDistObjectRel, RowExclusiveLock);
+}
+
+
+/*
+ * isObjectDistributed returns if the object addressed is already distributed in the
+ * cluster. This performs a local indexed lookup in pg_dist_object.
+ */
+bool
+isObjectDistributed(const ObjectAddress *address)
+{
+	Relation pgDistObjectRel = NULL;
+	ScanKeyData key[2] = { 0 };
+	SysScanDesc pgDistObjectScan = NULL;
+	HeapTuple pgDistObjectTup = NULL;
+	bool result = false;
+
+	pgDistObjectRel = heap_open(DistObjectRelationId(), AccessShareLock);
+
+	/* scan pg_dist_object for classid = $1 AND objid = $2 using an index */
+	ScanKeyInit(&key[0], Anum_pg_dist_object_classid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(address->classId));
+	ScanKeyInit(&key[1], Anum_pg_dist_object_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(address->objectId));
+	pgDistObjectScan = systable_beginscan(pgDistObjectRel,
+										  DistObjectClassIDObjectIDIndexId(), true, NULL,
+										  2, key);
+
+	while (HeapTupleIsValid(pgDistObjectTup = systable_getnext(pgDistObjectScan)))
+	{
+		/* tuple found, we are done */
+		result = true;
+		break;
+	}
+
+	systable_endscan(pgDistObjectScan);
+	relation_close(pgDistObjectRel, AccessShareLock);
+
+	return result;
+}
+
+
+/*
+ * citus_prepare_pg_upgrade_pg_dist_object prepares pg_dist_object before a pg_upgrade. It
+ * stores an upgrade stable identifier of the objects in the identifier column which can
+ * later be restored to the object id of the object.
+ */
+Datum
+citus_prepare_pg_upgrade_pg_dist_object(PG_FUNCTION_ARGS)
+{
+	Relation pgDistObject = NULL;
+	SysScanDesc scanDescriptor = NULL;
+	HeapTuple heapTuple = NULL;
+	HeapTuple newHeapTuple = NULL;
+	TupleDesc tupleDescriptor = NULL;
+
+	Datum values[Natts_pg_dist_object];
+	bool isnull[Natts_pg_dist_object];
+	bool replace[Natts_pg_dist_object];
+	Datum datumArray[Natts_pg_dist_object];
+
+	pgDistObject = heap_open(DistObjectRelationId(), RowExclusiveLock);
```

Comment:
Super unlikely but we wouldn't want anyone else to access `DistObjectRelationId ()` during this operation. So, in that sense `ExclusiveLock` seems good, which allows reads to continue if anyone is interested in reading the entries? Similar comment for `citus_finish_pg_upgrade_pg_dist_object `.

Reply:
these functions have been removed in favour of a simple sql statement with a smaller C footprint.